pkg/core/blockchain.go

```diff
@@ -677,36 +677,117 @@ func (bc *Blockchain) GetStateModule() blockchainer.StateRoot {
-	writeBuf := io.NewBufBinWriter()
-	appExecResults := make([]*state.AppExecResult, 0, 2+len(block.Transactions))
-	if err := cache.StoreAsBlock(block, writeBuf); err != nil {
-		return err
-	}
-	writeBuf.Reset()
+	var (
+		cache          = bc.dao.GetWrapped()
+		appExecResults = make([]*state.AppExecResult, 0, 2+len(block.Transactions))
+		aerchan        = make(chan *state.AppExecResult, len(block.Transactions)/8) // Tested 8 and 4 with no practical difference, but feel free to test more and tune.
+		aerdone        = make(chan error)
+		blockdone      = make(chan error)
+	)
+	go func() {
+		var (
+			kvcache  = cache.GetWrapped()
+			writeBuf = io.NewBufBinWriter()
+		)
+		if err := kvcache.StoreAsBlock(block, writeBuf); err != nil {
+			blockdone <- err
+			return
+		}
+		writeBuf.Reset()
 
-	if err := cache.StoreAsCurrentBlock(block, writeBuf); err != nil {
-		return err
-	}
-	writeBuf.Reset()
+		if err := kvcache.StoreAsCurrentBlock(block, writeBuf); err != nil {
+			blockdone <- err
+			return
+		}
+		writeBuf.Reset()
+
+		for _, tx := range block.Transactions {
+			if err := kvcache.StoreAsTransaction(tx, block.Index, writeBuf); err != nil {
+				blockdone <- err
+				return
+			}
 
+			writeBuf.Reset()
+			if bc.config.P2PSigExtensions {
+				for _, attr := range tx.GetAttributes(transaction.ConflictsT) {
+					hash := attr.Value.(*transaction.Conflicts).Hash
+					dummyTx := transaction.NewTrimmedTX(hash)
+					dummyTx.Version = transaction.DummyVersion
+					if err := kvcache.StoreAsTransaction(dummyTx, block.Index, writeBuf); err != nil {
+						blockdone <- fmt.Errorf("failed to store conflicting transaction %s for transaction %s: %w", hash.StringLE(), tx.Hash().StringLE(), err)
+						return
+					}
+					writeBuf.Reset()
+				}
+			}
+		}
+		if bc.config.RemoveUntraceableBlocks {
+			if block.Index > bc.config.MaxTraceableBlocks {
+				index := block.Index - bc.config.MaxTraceableBlocks // is at least 1
+				err := kvcache.DeleteBlock(bc.headerHashes[index], writeBuf)
+				if err != nil {
+					bc.log.Warn("error while removing old block",
+						zap.Uint32("index", index),
+						zap.Error(err))
+				}
+				writeBuf.Reset()
+			}
+		}
+		_, err := kvcache.Persist()
+		if err != nil {
+			blockdone <- err
+		}
+		close(blockdone)
+	}()
+	go func() {
+		var (
+			kvcache     = dao.NewCached(cache)
+			writeBuf    = io.NewBufBinWriter()
+			err         error
+			appendBlock bool
+		)
+		for aer := range aerchan {
+			if aer.Container == block.Hash() && appendBlock {
+				err = kvcache.AppendAppExecResult(aer, writeBuf)
+			} else {
+				err = kvcache.PutAppExecResult(aer, writeBuf)
+				if aer.Container == block.Hash() {
+					appendBlock = true
+				}
+			}
+			if aer.Execution.VMState == vm.HaltState {
+				for j := range aer.Execution.Events {
+					bc.handleNotification(&aer.Execution.Events[j], kvcache, block, aer.Container)
+				}
+			}
+			if err != nil {
+				err = fmt.Errorf("failed to store tx exec result: %w", err)
```

Comment:
Or block exec result. Maybe perform this error handling a bit earlier, before notifications handling?

Reply:
Yeah, makes sense.